### PR TITLE
Add OCB encryption mode

### PR DESCRIPTION
**Pull request by mtausig**
### Contribution description

This code implements the OCB(3) authenticated encryption mode according to RFC 7253.

### Testing procedure

Test vectors from the RFC have been used to create unit tests.

Just run `make tests-crypto` in the `tests/unittests` folder.

### CODE REVIEW

File: sys/crypto/modes/ocb.c [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_crypto
+ * @{
+ *
+ * @file        ocb.c
```
Comment (jia200x) [inline sys/crypto/modes/ocb.c]: alignment here ;)

File: sys/crypto/modes/ocb.c [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        cipher_input[remaining_data_len] = 0x80;
+        xor_block(cipher_input, offset, cipher_input);
+        /* Sum = Sum_m xor ENCIPHER(K, CipherInput) */
+        uint8_t enciphered_block[16];
+        cipher->interface->encrypt(&(cipher->context), cipher_input, enciphered_block);
+        xor_block(output, enciphered_block, output);
+    }
+}
+
+static int32_t run_ocb(cipher_t *cipher, uint8_t *auth_data, uint32_t auth_data_len,
```
Comment (jia200x) [inline sys/crypto/modes/ocb.c]: This functions is too big. Please split it so it's easier to maintain (as stated in [Functions @ Coding Conventions](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions#functions))

File: sys/crypto/modes/ocb.c [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        // We would not be able to return the proper output length for data this long
+        return OCB_ERR_INVALID_DATA_LENGTH;
+    }
+
+    int cipher_text_length = run_ocb(cipher, auth_data, auth_data_len,
+                                     tag, tag_len, nonce, nonce_len,
+                                     input, input_len, output, OCB_MODE_ENCRYPT);
+
+    if (cipher_text_length < 0) {
+        // An error occured. Retur the error code
```
Comment (jia200x) [inline sys/crypto/modes/ocb.c]: s/Retur/Return

File: sys/crypto/modes/ocb.c (hunk 1 above)
Comment (mtausig) [reply sys/crypto/modes/ocb.c]: What do you mean exactly? The file was processed with uncrustify, the alignment is as far as I can tell the same as in the other source files I checked.

File: sys/crypto/modes/ocb.c (hunk 3 above)
Comment (mtausig) [reply sys/crypto/modes/ocb.c]: e8cdaab33032cf1798fdebe6230c528e4368bae4

File: sys/crypto/modes/ocb.c (hunk 1 above)
Comment (jia200x) [reply sys/crypto/modes/ocb.c]: For some reason Uncrustify didn't align `ocb.c` with the next block (`Offset Codebook...`)

Comment (mtausig) [reply sys/crypto/modes/ocb.c]: Sorry, I still don't get what you mean. The block look perfectly aligend to me. 
The second column in line 13 ("ocb.c") starts at the very same position as the second column in the next line ("Offset Codebook (OCB3) AEAD mode as specified in RFC 7253")

Comment (gebart) [reply sys/crypto/modes/ocb.c]: It does not matter. We don't name the file name in the doxygen `@file ` special command. Use only `@file` with no argument.

File: sys/crypto/modes/ocb.c (hunk 2 above)
Comment (mtausig) [reply sys/crypto/modes/ocb.c]: I optimized it with cd1c167, d120d55 and 09ae9de.
The function is still a bit over one screen page, but every shortening from this point would only be an artificial one (copying blocks into separate functions which are only called once).
I hope it is OK for you this way.

File: sys/crypto/modes/ocb.c (hunk 1 above)
Comment (mtausig) [reply sys/crypto/modes/ocb.c]: OK? 
4933bb7

Comment (jia200x) [reply sys/crypto/modes/ocb.c]: @mtausig sorry, the web browser pulled a trick on me (even with raw mode!)

File: sys/crypto/modes/ocb.c (hunk 2 above)
Comment (jia200x) [reply sys/crypto/modes/ocb.c]: yes, it's OK. Thanks!

File: sys/crypto/modes/ocb.c (hunk 3 above)
Comment (jia200x) [reply sys/crypto/modes/ocb.c]: roger

File: sys/crypto/modes/ocb.c [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+    uint8_t cipher_output[16], cipher_input[16];
+    xor_block(input, state->offset, cipher_input);
+    if (mode == OCB_MODE_ENCRYPT) {
+        state->cipher->interface->encrypt(&(state->cipher->context), cipher_input, cipher_output);
+    }
+    else if (mode == OCB_MODE_DECRYPT) {
+        state->cipher->interface->decrypt(&(state->cipher->context), cipher_input, cipher_output);
+    }
+    xor_block(state->offset, cipher_output, output);
+    if (state->checksum != NULL) {
```
[the remark this answers is not in the document]
Comment (mtausig) [reply sys/crypto/modes/ocb.c]: Fixed with 57e9e0c

**Comment by mtausig** [comment]
Squashed & Rebased

**Comment by mtausig** [comment]
Ping

**Comment by mtausig** [comment]
Ping

**Comment by jia200x** [comment]
I confirm all tests pass OK (native and samr21-xpro).
I'm fine codewise.

Could you split the commits in 2? (one for the implementation and one for the tests). With that, I'm fine to give the ACK

**Comment by mtausig** [comment]
@jia200x Done.

**Comment by jia200x** [comment]
thanks!

**Pull request merged**